pkg/instancetype/compatibility/compatibility.go

```diff
@@ -0,0 +1,196 @@
+//nolint:dupl,lll,gocyclo
+package compatibility
+
+import (
+	"fmt"
+
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/json"
+
+	"kubevirt.io/api/instancetype/v1alpha1"
+	"kubevirt.io/api/instancetype/v1alpha2"
+	"kubevirt.io/api/instancetype/v1beta1"
+	generatedscheme "kubevirt.io/client-go/generated/kubevirt/clientset/versioned/scheme"
+)
+
+func GetInstancetypeSpec(revision *appsv1.ControllerRevision) (*v1beta1.VirtualMachineInstancetypeSpec, error) {
+	if err := Decode(revision); err != nil {
+		return nil, err
+	}
+	switch obj := revision.Data.Object.(type) {
+	case *v1beta1.VirtualMachineInstancetype:
+		return &obj.Spec, nil
+	case *v1beta1.VirtualMachineClusterInstancetype:
+		return &obj.Spec, nil
+	default:
+		return nil, fmt.Errorf("unexpected type in ControllerRevision: %T", obj)
+	}
+}
+
+func GetPreferenceSpec(revision *appsv1.ControllerRevision) (*v1beta1.VirtualMachinePreferenceSpec, error) {
+	if err := Decode(revision); err != nil {
+		return nil, err
+	}
+	switch obj := revision.Data.Object.(type) {
+	case *v1beta1.VirtualMachinePreference:
+		return &obj.Spec, nil
+	case *v1beta1.VirtualMachineClusterPreference:
+		return &obj.Spec, nil
+	default:
+		return nil, fmt.Errorf("unexpected type in ControllerRevision: %T", obj)
+	}
+}
+
+func Decode(revision *appsv1.ControllerRevision) error {
+	if len(revision.Data.Raw) == 0 {
```

Comment:
is there a meaning that this is empty? should we log anything?

Reply:
Yeah we need to handle this better, I think this was originally put in place for the tests as it's always populated in runtime `ControllerRevisions`. I can follow up.